### ISSUE TITLE
ability to write methods that would be available for all components

### ISSUE THREAD

**Issue by maxdee** [praxis-live/support#117]
You probably want to rename this one...
It would be nice if there was a easy way to add code that could be used by any components. Lets say a user has a method they like to use in many components, instead of copy pasting it in all the components, the code could live in a single place but be accessed by many components.

In my recent patch I was rendering scenes smaller then the width/height and then tiled them together. I had a method I copy pasted in the rendering components which would define a different width/height. If I wanted to change this I had to close the project and do a search/replace on the .pxg file. It was an annoying work around.

**Comment by neilcsmith-net** [comment]
This is possible already, even if the API is a bit awkward for now, using Data ports and sharing functions. See this project, particularly the audio patch - https://github.com/neilcsmith-net/praxislive-iclc2019/tree/master/Projects/DataPipeTesting

This can/should be improved!

**Comment by maxdee** [comment]
Oh right I did see that, but I would prefer to do this without having to add more wires... For example I would have a `line(Pvector a, PVector b)` defined somewhere and accessible by all P3D components, without having to add a wire.

**Comment by neilcsmith-net** [comment]
Yes, it's a side effect of the Data pipe support that I've been experimenting with. But it would be good to support without connections (or just hide connections by category?) One pro of this approach is that we can support composite functions. 

Now we can't directly support method calls this way. We need to use function interfaces. It's probably unlikely we can ever support method calls.

The other problem with supporting your example is that you'd have to pass in the graphics context to draw on too somehow.

**Comment by neilcsmith-net** [comment]
Been thinking about this a lot again!  Better support for code re-use is high on the agenda, but not easy to get right.  The key thing is that all nodes are encapsulated - there is no way for one node to reference the actual Class or instance of another.  We cannot break that contract without breaking the ability to do safe hot code reloading at all.

It might be possible to use something like `LambdaMetaFactory` or `MethodHandle` along with some automatic code generation to inject something that makes this work.  As long as we do not cache between calls, and handle the possibility of the other code disappearing (both of which get handled in the Data pipe example)

It would be nice not to end up with the `line(PVector a, PVector b)` example having to be coded by the user as this, even if it's close to what is happening :smile: -

```
Optional<BiConsumer<PVector, PVector>> line = ...;
line.ifPresent(l -> l.accept(a, b));
```

**Comment by maxdee** [comment]
Maybe a library like approach would be better? These helper functions don't really need to be live codeable. Making a Processing library has been on my to learn list for a long time (and coding train just made a tutorial).

**Comment by neilcsmith-net** [comment]
@maxdee well, Java (and Processing) libraries can already be added to a project.  Although there's no support for compiling them in PraxisLIVE itself, it might be something we could provide basic support for.  But I'd still rather solve this properly in a way that is in the spirit of PraxisCORE / PraxisLIVE - everything should be hot-swappable at runtime.  That's not just about live coding in a performative sense.

**Comment by maxdee** [comment]
I entirely agree.

**Comment by neilcsmith-net** [comment]
Been thinking a little more about this but not getting too far! :confounded: The problem is handling the method changing, changing signature, or disappearing completely.  One option might be using MethodHandle very directly - something like.

/video/utils component

```
@Export void line(Pvector a, PVector b) { .... }
```

/video/*** component

```
@Import("utils::line") MethodHandle line;

public void draw() {
  line.invoke(a, b);
}
```

Except it'll probably need try / catching.  Other way may be an interface, something like ..

```
interface Utils {
  void line(PVector a, PVector b);
}
@Import("utils") Utils u;

public void draw() {
  u.line(a, b);
}
```

**Comment by neilcsmith-net** [comment]
I'm also considering the possibility of shared classes being available via the container. These would be put in a package called `SHARED`. Any component that had `import SHARED.<classname>` would be automatically reloaded whenever any of the shared classes was rewritten. This should allow for both shared code and shared types (data classes) to be passed around a graph.

**Comment by neilcsmith-net** [comment]
Finally working on the `SHARED` code support - basic principle now working in branch at https://github.com/praxis-live/praxiscore/tree/code-revisions - needs some polishing and editor support, but will hopefully make it in in some fashion to v5.2.

**Comment by neilcsmith-net** [comment]
Merged to master and will be available in v5.2.
